### PR TITLE
Feat/4095 4134  user space advanced search

### PR DESCRIPTION
**Pull request by asf-algoo**
See #4095 and #4134.

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [x] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified
- [x] Original authors of the features included in a multi-feature branch (maintenance fixes -> develop, security fixes -> develop, …) should be part of the reviewers, especially if you encountered merge conflicts.

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [ ] Manual, quality tests have been done

### CODE REVIEW

**Review by raphj — APPROVED** [review]

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 1]
```diff
@@ -536,6 +627,143 @@ def search_content(
...
+        fields = [
+            "public_name.exact^5",
+            "username.exact^5",
+            "public_name^3",
+            "username^3",
+            "custom_properties",
+        ]
+
+        search_fields = search_fields or DEFAULT_USER_SEARCH_FIELDS
+        fields = [field for field in fields if name_starts_with_any_prefix(field, search_fields)]
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:656]: I do suggest to follow similar solution as content for this, using enum:
https://github.com/tracim/tracim/pull/4198/commits/c7622b309b6e072d4d53173cde97917b3066a492
I don't have strong opinion if we should fix this now or later, but i expect we try to have some consistency for this.

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 2]
```diff
@@ -536,6 +627,143 @@ def search_content(
...
+            index=self._get_index_parameters(IndexedWorkspace).alias,
+        )
+
+        fields = [
+            "label.exact^5",
+            "label^3",
+            "description^3",
+        ]
+        search_fields = search_fields or DEFAULT_WORKSPACE_SEARCH_FIELDS
+        fields = [field for field in fields if name_starts_with_any_prefix(field, search_fields)]
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:724]: similar advice: should be consistent with content sooner or later.

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 3]
```diff
@@ -281,6 +305,73 @@ def index_contents(self, contents: typing.Iterable[Content]) -> None:
...
+        )
+        indexed_user.meta.id = user.user_id
+        user_index_alias = self._get_index_parameters(self.IndexedUser).alias
+        indexed_user.save(
+            using=self.es,
+            index=user_index_alias,
+            request_timeout=self._config.SEARCH__ELASTICSEARCH__REQUEST_TIMEOUT,
+        )
+
+    def delete_user(self, user: User) -> None:
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:342]: Not directly related, but as we do "hard" deletion in elasticsearch here for user/workspace, how the deletion of content is handled ? softly/hardly/missing ?

File: backend/tracim_backend/views/search_api/schemas.py [hunk 4]
```diff
@@ -25,6 +27,17 @@ class SearchContentField(Enum):
     DESCRIPTION = "description"
 
 
+class UserSearchField(str, Enum):
```
Comment (inkhey) [inline backend/tracim_backend/views/search_api/schemas.py:30]: Humm i do probably miss something between this and default list before.

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py (hunk 3 above)
Comment (inkhey) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:342]: After discussion with @grignards soft deletion seems properly handled.

We should probably verify hard deletion too (using tracimcli command, when deleting user).

Comment (grignards) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:342]: The `is_deleted` case is handled in `on_content_modified`. There is currently no management for database-level delete of contents.

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 5]
```diff
@@ -43,6 +55,16 @@
 FILE_PIPELINE_DESTINATION_FIELD = "file_data"
 FILE_PIPELINE_LANGS = ["en", "fr", "pt", "de"]
 
+DEFAULT_CONTENT_SEARCH_FIELDS = [field for field in ContentSearchField]
+DEFAULT_USER_SEARCH_FIELDS = [field for field in UserSearchField]
+DEFAULT_WORKSPACE_SEARCH_FIELDS = [field for field in WorkspaceSearchField]
```
Comment (raphj) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:60]:
```suggestion
DEFAULT_CONTENT_SEARCH_FIELDS = list(ContentSearchField)
DEFAULT_USER_SEARCH_FIELDS = list(UserSearchField)
DEFAULT_WORKSPACE_SEARCH_FIELDS = list(WorkspaceSearchField)
```

**Review by raphj — APPROVED** [review]

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py (hunk 1 above)
Comment (grignards) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:659]: Done.

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py (hunk 2 above)
Comment (grignards) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:729]: Done.

**Pull request merged**